### DISCUSSION TITLE
Model viewer crash/reload on iPhone devices

### DISCUSSION THREAD

**Discussion by NitinJ88** [google/model-viewer#3740] [Q&A]
### Description
Model viewer is crashing or browser is reloading while model is loading in viewer on iPhone devices.
iPhone models:  8, 12, and 13

Code snippet:
![image](https://user-images.githubusercontent.com/86343802/186723549-0228afec-f164-42b5-847a-3f27ac6deb7f.png)

Console errors:
USDZExporter: Normals missing.
chunk-vendors.105ffd4a.js:1020 USDZExporter: UVs missing.

Please provide a detailed description of the bug, how to reproduce it, and the
expected behavior.  Always include a code snippet, screenshot, any errors
reported in the console, and the model to help us understand and fix the
problem.

#### Live Demo

https://user-images.githubusercontent.com/86343802/186728738-bedcfab1-a8bf-476c-ab70-fa5352ccbfe8.mp4

https://glitch.com/edit/#!/model-viewer

### Version

- model-viewer: v1.10.1

### Browser Affected

- [ ] Chrome, version: xx.x.xxxx.xx
- [ ] Edge
- [ ] Firefox
- [ ] IE
- [* ] Safari

### OS

- [ ] Android
- [*] iOS
- [ ] Linux
- [ ] MacOS
- [ ] Windows

### AR

- [*] WebXR
- [ ] SceneViewer
- [ ] QuickLook

**Comment by elalish** [comment]
This looks like an out-of-memory situation. Can you link to the page or the model so we can try to repro? You console logs about USDZExporter only happen when you try to enter AR, which doesn't seem to be part of your video (unless I've missed something?).

**Reply by NitinJ88** [reply]
This is one of the model failing on iPhone 8 with same reason. I am checking if can share the model showing on video. 
[astrolab 2.zip](https://github.com/google/model-viewer/files/9431925/astrolab.2.zip)

**Reply by elalish** [reply]
So, your model isn't huge, but it is pretty poorly-optimized for rendering; I'd guess you're hitting some kind of edge case in the iPhone's GPU, but you'll get much better results if you spend a little more time on this model anyway. Looking at "Model Details" in our [editor](https://modelviewer.dev/editor/) shows that it has 457 draw calls - this will kill performance. You'll want to bake this all down to a single material with a texture instead. Your model also doesn't include normal vectors, which is why it looks faceted; if you include those it'll look nice and smooth (this is also what USDZExporter is complaining about). Finally, you'll want an artist to create proper metal-rough PBR materials for you so this looks realistic.

**Reply by NitinJ88** [reply]
We use obj2gltf npm package to convert our model to glb. So we really don't have control to optimize the glb. Can you suggest any other alternative package? This issue was not seen on model viewer version 1.2.1. This issue started when we upgrade model viewer version to 1.10.1 where ios-src prop is optional.

**Reply by elalish** [reply]
I'd recommend Blender - it's free, always has the most up-to-date glTF export support, and has many tools for fixing models.

**Reply by NitinJ88** [reply]
We will explore the Blender option but it will take some time. Do we have any other solution and workaround for the issue?

**Reply by elalish** [reply]
Sorry no, I think it was only luck that this file ever rendered on this device. Exporting a well-formed, web-optimized GLB is pretty critical to get a good experience across devices.

**Reply by lvsonawane** [reply]
> I'd recommend Blender - it's free, always has the most up-to-date glTF export support, and has many tools for fixing models.

I tried with Blender tool to export obj to glTF/glb so I found that exported file has missing color information & its design not able to move into 360°. And file size also increased as compare to original glb file. Also trying to find blender npm package but i couldn`t get any appropriate npm package for it.

Can you please suggest npm package for blender or any another way to resolve this issue?

**Reply by elalish** [reply]
Blender is a desktop application, it has nothing to do with npm. I recommended it because you will need to re-author your OBJ; no simple converter will do the job as you have no PBR material properties yet. Find a 3D artist to fix it up for you.

**Reply by NitinJ88** [reply]
We have updated the latest model-viewer version 2.1.1 but still this issue is reproducing. One change in behavior is that model is now rendering on model-viewer but when I try to interact to model then model-viewer crashes.

**Reply by elalish** [reply]
The latest version is v3.0.2, but this thread says the model is the problem. Have you fixed that yet, or tried any other GLBs?

**Comment by mudworker** [comment]
I encountered the same problem (v3.3.0). Using model-viewer to load glb will cause the page to reload or crash directly.
It only appears on the mobile phone ( iphone safari, and not tested on Android yet ), and can be loaded successfully on the computer.
This is most likely caused by the model being too large (because it can be loaded successfully when loading a small model). Does model-viewer have strict requirements on model size?

**Reply by elalish** [reply]
It's not a MV requirement, but a GPU RAM limit, which varies by device (and other applications open). Try KTX2 textures to reduce your GPU RAM footprint.

**Reply by shawnadelic** [reply]
Would it make a difference if the crashing was only happening on iOS Safari (not iOS Chrome)? Or could that still point to a GPU RAM limit issue?

**Reply by mudworker** [reply]
it will also crashing on IOS Chrome. All browsers on ios are packaged based on safari. and it can be loaded normally on computer
